core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackend.scala

```diff
@@ -239,44 +250,69 @@ private[spark] class MesosSchedulerBackend(
+      val mem = getResource(o.getResourcesList, "mem")
+      val cpus = getResource(o.getResourcesList, "cpus")
+      val slaveId = o.getSlaveId.getValue
+      val offerAttributes = toAttributeMap(o.getAttributesList)
+
+      // check offers for
+      //  1. Memory requirements
+      //  2. CPU requirements - need at least 1 for executor, 1 for task
+      val meetsMemoryRequirements = mem >= calculateTotalMemory(sc)
+      val meetsCPURequirements = cpus >= (mesosExecutorCores + scheduler.CPUS_PER_TASK)
+      val meetsRequirements =
+        (meetsMemoryRequirements && meetsCPURequirements) ||
           (slaveIdToExecutorInfo.contains(slaveId) && cpus >= scheduler.CPUS_PER_TASK)
-        val debugstr = if (meetsRequirements) "Accepting" else "Declining"
-        logDebug(s"$debugstr offer: ${o.getId.getValue} with attributes: "
-          + s"$offerAttributes mem: $mem cpu: $cpus")
+      val debugstr = if (meetsRequirements) "Accepting" else "Declining"
+      logDebug(s"$debugstr offer: ${o.getId.getValue} with attributes: "
+ 				+ s"$offerAttributes mem: $mem cpu: $cpus")
+
+      meetsRequirements
+    }
+
+    // Decline offers we ruled out immediately
+    unUsableOffers.foreach(o => d.declineOffer(o.getId))
+
+    var availableCores = Math.max(0, maxCores - totalCoresAcquired)
 
-        meetsRequirements
+    val workerOffers = (for (o <- usableOffers) yield {
+      val coresInOffer = getResource(o.getResourcesList, "cpus").toInt
+      val extraCores = if (slaveIdToExecutorInfo.contains(o.getSlaveId.getValue)) {
+        0D
+      } else {
+        // If the Mesos executor has not been started on this slave yet, set aside a few
+        // cores for the Mesos executor by offering fewer cores to the Spark executor
+        mesosExecutorCores
       }
 
-      // Decline offers we ruled out immediately
-      unUsableOffers.foreach(o => d.declineOffer(o.getId))
+      // the cores we can offer for tasks on workers should not exceed neither availableCores
+      // nor cores in the current offer, after accounting for non-task cores
+      val taskCores = Math.min(availableCores - extraCores, coresInOffer - extraCores)
 
-      val workerOffers = usableOffers.map { o =>
-        val cpus = if (slaveIdToExecutorInfo.contains(o.getSlaveId.getValue)) {
-          getResource(o.getResourcesList, "cpus").toInt
-        } else {
-          // If the Mesos executor has not been started on this slave yet, set aside a few
-          // cores for the Mesos executor by offering fewer cores to the Spark executor
-          (getResource(o.getResourcesList, "cpus") - mesosExecutorCores).toInt
-        }
-        new WorkerOffer(
+      if (taskCores > 0) {
+        availableCores -= taskCores + extraCores
+        Option(new WorkerOffer(
           o.getSlaveId.getValue,
           o.getHostname,
-          cpus)
+          taskCores.toInt))
+      } else {
+        None
       }
+    }).flatten
+
+    (usableOffers, workerOffers)
```

Comment:
What happens if we hit the limit and don't run any more tasks on the remaining usableOffers? We need to decline those too right?